infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/runtime/LogReplicationClientServerRouter.java

```diff
@@ -0,0 +1,605 @@
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.concurrent.atomic.AtomicLong;
+
+import static org.corfudb.protocols.CorfuProtocolCommon.getUuidMsg;
+import static org.corfudb.protocols.service.CorfuProtocolMessage.getDefaultProtocolVersionMsg;
+import static org.corfudb.protocols.service.CorfuProtocolMessage.getRequestMsg;
+import static org.corfudb.protocols.service.CorfuProtocolMessage.getResponseMsg;
+
+/**
+ * Router, interfaces between the custom transport layer and LR core components.
+ *
+ * There is one router per node, and all the sessions (both incoming and outgoing) share the router.
+ */
+@Slf4j
+public class LogReplicationClientServerRouter implements IClientServerRouter {
+
+    public static final String REMOTE_LEADER = "REMOTE_LEADER";
+
+    /**
+     * A map of session to {@link CompletableFuture}, which is completed when a connection,
+     * including a successful handshake, completes and messages can be sent
+     * to the remote node.
+     */
+    @Getter
+    private volatile Map<LogReplicationSession, CompletableFuture<Void>> sessionToLeaderConnectionFuture;
+
+    /**
+     * A map of session to the current requestID for the session.
+     */
+    @Getter
+    @SuppressWarnings("checkstyle:abbreviation")
+    private final Map<LogReplicationSession,AtomicLong> sessionToRequestIdCounter;
+
+    /**
+     * Sync call response timeout (milliseconds).
+     */
+    @Getter
+    @Setter
+    private long timeoutResponse;
+
+    /**
+     * The outstanding requests on this router.
+     */
+    @Getter
+    private final Map<LogReplicationSession, Map<Long, CompletableFuture>> sessionToOutstandingRequests;
+
+    /**
+     * Runtime FSM, to insert connectivity events
+     */
+    @Getter
+    private final Map<LogReplicationSession, CorfuLogReplicationRuntime> sessionToRuntimeFSM;
+
+    /**
+     * Remote cluster's clusterDescriptor
+     */
+    @Getter
+    private final Map<LogReplicationSession, ClusterDescriptor> sessionToRemoteClusterDescriptor;
+
+    /**
+     * The replicationManager
+     */
+    private final CorfuReplicationManager replicationManager;
+
+    /**
+     * local cluster ID
+     */
+    private String localClusterId;
+
+    /**
+     * Client Transport adapter
+     */
+    @Setter
+    private IClientChannelAdapter clientChannelAdapter;
+
+    /**
+     * Server transport adapter
+     */
+    @Setter
+    @Getter
+    private IServerChannelAdapter serverChannelAdapter;
+
+    /**
+     * Log replication server which has the handlers for incoming messages.
+     */
+    @Getter
+    private final LogReplicationServer msgHandler;
+
+    /**
+     * Triggers leadership request and create bidirectional streams to remote endpoints
+     */
+    private final Map<LogReplicationSession, VerifyRemoteSourceLeader> sessionToSinkVerifyLeadership;
```

Comment:
sessionToSource?

Reply:
No, since this one is a map of session -> verifyLeadership object which is per session.

I'll be reevaluating this to having the leadershipQuery go out just once per remote cluster. But that will be in a followup PR